ocaml/xapi/xapi_vdi.ml

```diff
@@ -550,11 +578,93 @@ let snapshot ~__context ~vdi ~driver_params =
   update_allowed_operations ~__context ~self:newvdi;
   newvdi
 
+(** Wait for all the VBDs of this VDI to be unplugged and destroyed.
+    It is sufficient to wait for all the VBDs to disappear, because a VBD
+    cannot be destroyed if it is plugged. *)
+let wait_for_vbds_to_be_unplugged_and_destroyed ~__context ~self =
+  let vdi_uuid = Db.VDI.get_uuid ~__context ~self in
+  (* We watch the "VBDs" field of this VDI. *)
+  let classes = [Printf.sprintf "VDI/%s" (Ref.string_of self)] in
+
+  let next_token_and_vbds ~token ~timeout =
+    let most_recent_vbds_field events =
+      (* We do not assume anything here about the order of the list of events we get. *)
+      let most_recent_snapshot =
+        let events_from_newest_to_oldest =
+          (* We need to sort the timestamp strings in decreasing order *)
+          List.sort (fun e1 e2 -> Event_types.(- String.compare e1.ts e2.ts)) events
+        in
+        let snapshots_from_newest_to_oldest =
+          (* filter_map in Stdext preserves the order of elements *)
+          Xapi_stdext_std.Listext.List.filter_map
+            (fun event -> event.Event_types.snapshot)
+            events_from_newest_to_oldest
+        in
+        Xapi_stdext_std.Listext.List.safe_hd snapshots_from_newest_to_oldest
+      in
+      Xapi_stdext_monadic.Opt.map
+        (fun snapshot ->
+           let vdi = API.vDI_t_of_rpc snapshot in
+           vdi.API.vDI_VBDs)
+        most_recent_snapshot
+    in
+
+    let from =
+      Helpers.call_api_functions ~__context
+        (fun rpc session_id ->
+           Client.Event.from ~rpc ~session_id ~classes ~token ~timeout |> Event_types.event_from_of_rpc)
+        ~test_fn:(fun () -> Xapi_event.from ~__context ~classes ~token ~timeout |> Event_types.parse_event_from)
```

Comment:
We'd better not set `Helpers.test_mode` to true in production, cause then data_destroy, and possibly other functions will likely break.